### PR TITLE
Dto and services

### PR DESCRIPTION
**Pull request by MalishkaMyu**
In the begin of the week ich tried to create pull request but probably made something wrong. One more try

### CODE REVIEW

File: src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeDTOToEntityConverter.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.samsolutions.employeesdep.model.entities.Employee;
+import org.springframework.beans.BeanUtils;
+import org.springframework.core.convert.converter.Converter;
+
+public class EmployeeDTOToEntityConverter implements Converter<EmployeeDTO, Employee> {
+    @Override
+    public Employee convert(EmployeeDTO source) {
+        Employee target = new Employee();
+        BeanUtils.copyProperties(source,target);
+        //target.setDepartment(new DepartmentDTOToEntityConverter().convert(source.getDepartment()));
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeDTOToEntityConverter.java:13]: commented code should be removed (comments should not be in a main branch)

File: src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeEntityToDTOConverter.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+
+public class EmployeeEntityToDTOConverter implements Converter<Employee, EmployeeDTO> {
+    @Override
+    public EmployeeDTO convert(Employee source) {
+        EmployeeDTO target = new EmployeeDTO();
+        BeanUtils.copyProperties(source, target);
+        target.setWorkExperience(ChronoUnit.YEARS.between(target.getEmploymentDate(), LocalDate.now()));
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeEntityToDTOConverter.java:20]: workExperience is a strange field. If we have employment date and work experience is always calculated - maybe we should remove this field from the database completely?

File: src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java [hunk 3]
```diff
@@ -1,42 +1,20 @@
...
-    public Employee saveEmployee(Employee employeeToSave) {
-        Department departToSave = employeeToSave.getDepartment();
-        departRepository.save(departToSave);
-        for (Role role:employeeToSave.getEmployeeRoles()) {
-            roleRepository.save(role);
-        }
-        return empRepository.save(employeeToSave);
-    }
+public interface EmployeeService {
+    List<EmployeeDTO> getEmployeesToDepartment(String departName);
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/EmployeeService.java:9]: add empty line before

File: src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+import javax.transaction.Transactional;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class EmployeeServiceImpl implements EmployeeService {
+    @Value("${employees.per.page}")
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java:30]: add empty line before

File: src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        // saving employee
+        empRepository.save(employeeToSave);
+        // convert employee entity back to DTO to have all IDs
+        return new EmployeeEntityToDTOConverter().convert(employeeToSave);
+    }
+
+    @Override
+    @Transactional
+    public void deleteEmployeeById(Long employeeId) {
+        if (empRepository.existsById(employeeId))
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java:126]: add square brackets

File: src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            empRepository.deleteById(employeeId); // how to know whether delete was successful???
```
Comment (AlecKotovichSAM) [inline src/main/java/com/samsolutions/employeesdep/model/services/impl/EmployeeServiceImpl.java:127]: how to know whether delete was successful??? - you may search for it after removal and check it is removed

File: src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    private JpaRoleDao roleRepository;
+
+    private final List<Long> listIDs = new ArrayList<>();
+
+    private void updateRoles(RoleDTO[] allRoles, Set<RoleDTO> savedRoles) {
+        for (RoleDTO savedRole : savedRoles) {
+            for (RoleDTO role : allRoles) {
+                if (role.getRole().equals(savedRole.getRole())) {
+                    role.setId(savedRole.getId());
+                    break;
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java:41]: did not understand this method

File: src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @BeforeEach
+    public void setup() {
+        EmployeeDTO empToSave, savedEmp;
+        DepartmentDTO depart1 = new DepartmentDTO("Java Department");
+        DepartmentDTO depart2 = new DepartmentDTO(".NET Department");
+        RoleDTO[] allRoles = {new RoleDTO("Java Programmer"), new RoleDTO(".NET programmer"),
+                new RoleDTO("Tester"), new RoleDTO("Project manager"), new RoleDTO("coffee drinker")};
+        // creating and saving employee 1
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java:54]: add empty line before

File: src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @Test
+    public void findAllEmployeesWithPages() {
+        // first page
+        List<EmployeeDTO> employees = empService.getAllEmployees();
+        assertEquals(2, employees.size());
+        assertEquals("Krosh", employees.get(0).getName());
+        assertEquals("Nyusha", employees.get(1).getName());
+        // second page
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java:108]: add empty line before

File: src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        employees = empService.getAllEmployees(1);
+        assertEquals(1, employees.size());
+        assertEquals("Pin", employees.get(0).getName());
+    }
+
+    @AfterEach
+    public void tearDown() {
+        for (Long id : listIDs)
```
Comment (AlecKotovichSAM) [inline src/test/java/com/samsolutions/employeesdep/services/EmployeeServiceTest.java:116]: add square brackets

**Review by AlecKotovichSAM — CHANGES_REQUESTED** [review]

File: src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeEntityToDTOConverter.java (hunk 2 above)
Comment (MalishkaMyu) [reply src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeEntityToDTOConverter.java:20]: the field workExperience is not a DB field, it presents only in DTO

Comment (AlecKotovichSAM) [reply src/main/java/com/samsolutions/employeesdep/model/converters/EmployeeEntityToDTOConverter.java:20]: ok

**Review by AlecKotovichSAM — APPROVED** [review]

**Pull request merged**